pilot/pkg/serviceregistry/serviceentry/controller.go

```diff
@@ -865,30 +871,18 @@ func servicesDiff(os []*model.Service, ns []*model.Service) ([]*model.Service, [
 // allocated IP addresses do not take effect.
 //
 // The current algorithm to allocate IPs is deterministic across all istiods.
-// At stable state, given two istiods with exact same set of services, there should
-// be no change in XDS as the algorithm is just a dumb iterative one that allocates sequentially.
-//
-// TODO: Rather than sequentially allocate IPs, switch to a hash based allocation mechanism so that
-// deletion of the oldest service entry does not cause change of IPs for all other service entries.
-// Currently, the sequential allocation will result in unnecessary XDS reloads (lds/rds) when a
-// service entry with auto allocated IP is deleted. We are trading off a perf problem (xds reload)
-// for a usability problem (e.g., multiple cloud SQL or AWS RDS tcp services with no VIPs end up having
-// the same port, causing traffic to go to the wrong place). Once we move to a deterministic hash-based
-// allocation with deterministic collision resolution, the perf problem will go away. If the collision guarantee
-// cannot be made within the IP address space we have (which is about 64K services), then we may need to
-// have the sequential allocation algorithm as a fallback when too many collisions take place.
 func autoAllocateIPs(services []*model.Service) []*model.Service {
-	// i is everything from 240.240.0.(j) to 240.240.255.(j)
-	// j is everything from 240.240.(i).1 to 240.240.(i).254
-	// we can capture this in one integer variable.
-	// given X, we can compute i by X/255, and j is X%255
-	// To avoid allocating 240.240.(i).255, if X % 255 is 0, increment X.
-	// For example, when X=510, the resulting IP would be 240.240.2.0 (invalid)
-	// So we bump X to 511, so that the resulting IP is 240.240.2.1
-	maxIPs := 255 * 255 // are we going to exceed this limit by processing 64K services?
-	x := 0
-	hnMap := make(map[string]octetPair)
-
+	hashedServices := make([]*model.Service, maxIPs)
+	hash := fnv.New32a()
+	// First iterate through the range of services and determine its position by hash
+	// so that we can deterministically allocate an IP.
+	// We use "Double Hashning" for collision detection.
+	// The hash algorithm is
+	// - h1(k) = Sum32 hash of the host name.
+	// - Check if we have an empty slot for h1(x) % MAXIPS. Use it if available.
+	// - If there is a collision, apply second hash i.e. h2(x) = PRIME - (Key % PRIME)
+	//   where PRIME is the max prime number below MAXIPS.
+	// - Calculate new hash iteratively till we find an empty slot with (h1(k) + i*h2(k)) % MAXIPS
 	for _, svc := range services {
 		// we can allocate IPs only if
```

Comment:
Are services sorted ? So we don't flip flop on different Istiod instances ? 

Setting status or adding an annotation would avoid most of the conflict problems, and if we delve into IPv6 allocation we can also replace hashing with a structured format, for example allocate prefixes to each namespace (and even service account ).

Reply:
Yes. They are sorted